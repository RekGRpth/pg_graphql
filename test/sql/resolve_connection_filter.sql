begin;
    create table account(
        id int primary key,
        is_verified bool
    );

    insert into public.account(id, is_verified)
    values
        (1, true),
        (2, true),
        (3, false);

    savepoint a;

    -- Filter by Int
    select jsonb_pretty(
        graphql.resolve($$
            {
              accountCollection(filter: {id: {eq: 2}}) {
                edges {
                  node {
                    id
                  }
                }
              }
            }
        $$)
    );
    rollback to savepoint a;

    -- Filter by Int and bool. has match
    select jsonb_pretty(
        graphql.resolve($$
            {
              accountCollection(filter: {id: {eq: 2}, isVerified: {eq: true}}) {
                edges {
                  node {
                    id
                  }
                }
              }
            }
        $$)
    );
    rollback to savepoint a;

    -- Filter by Int and bool. no match
    select jsonb_pretty(
        graphql.resolve($$
            {
              accountCollection(filter: {id: {eq: 2}, isVerified: {eq: false}}) {
                edges {
                  node {
                    id
                  }
                }
              }
            }
        $$)
    );
    rollback to savepoint a;


    -- Variable: Int
    select jsonb_pretty(
        graphql.resolve($$
           query AccountsFiltered($filt: Int!)
           {
             accountCollection(filter: {id: {eq: $filt}}) {
               edges {
                 node{
                   id
                 }
               }
             }
           }
        $$,
        variables:= '{"filt": 1}'
      )
    );
    rollback to savepoint a;

    -- Variable: IntFilter
    select jsonb_pretty(
        graphql.resolve($$
           query AccountsFiltered($ifilt: IntFilter!)
           {
             accountCollection(filter: {id: $ifilt}) {
               edges {
                 node{
                   id
                 }
               }
             }
           }
        $$,
        variables:= '{"ifilt": {"eq": 3}}'
      )
    );
    rollback to savepoint a;

    -- Variable: AccountFilter, single col
    select jsonb_pretty(
        graphql.resolve($$
           query AccountsFiltered($afilt: AccountFilter!)
           {
             accountCollection(filter: $afilt) {
               edges {
                 node{
                   id
                 }
               }
             }
           }
        $$,
        variables:= '{"afilt": {"id": {"eq": 2}} }'
      )
    );
    rollback to savepoint a;

    -- Variable: AccountFilter, multi col match
    select jsonb_pretty(
        graphql.resolve($$
           query AccountsFiltered($afilt: AccountFilter!)
           {
             accountCollection(filter: $afilt) {
               edges {
                 node{
                   id
                 }
               }
             }
           }
        $$,
        variables:= '{"afilt": {"id": {"eq": 2}, "isVerified": {"eq": true}} }'
      )
    );
    rollback to savepoint a;

    -- Variable: AccountFilter, multi col no match
    select jsonb_pretty(
        graphql.resolve($$
           query AccountsFiltered($afilt: AccountFilter!)
           {
             accountCollection(filter: $afilt) {
               edges {
                 node{
                   id
                 }
               }
             }
           }
        $$,
        variables:= '{"afilt": {"id": {"eq": 2}, "isVerified": {"eq": false}} }'
      )
    );
    rollback to savepoint a;

    -- Variable: AccountFilter, invalid field name
    select jsonb_pretty(
        graphql.resolve($$
           query AccountsFiltered($afilt: AccountFilter!)
           {
             accountCollection(filter: $afilt) {
               edges {
                 node{
                   id
                 }
               }
             }
           }
        $$,
        variables:= '{"afilt": {"dne_id": 2} }'
      )
    );
    rollback to savepoint a;

    -- Variable: AccountFilter, invalid IntFilter
    select jsonb_pretty(
        graphql.resolve($$
           query AccountsFiltered($afilt: AccountFilter!)
           {
             accountCollection(filter: $afilt) {
               edges {
                 node{
                   id
                 }
               }
             }
           }
        $$,
        variables:= '{"afilt": {"id": 2} }'
      )
    );
    rollback to savepoint a;

    -- Variable: AccountFilter, invalid data type
    select jsonb_pretty(
        graphql.resolve($$
           query AccountsFiltered($afilt: AccountFilter!)
           {
             accountCollection(filter: $afilt) {
               edges {
                 node{
                   id
                 }
               }
             }
           }
        $$,
        variables:= '{"afilt": {"id": {"eq": true}} }'
      )
    );
    rollback to savepoint a;

rollback;
