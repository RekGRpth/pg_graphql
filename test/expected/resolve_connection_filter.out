begin;
    create table account(
        id int primary key,
        is_verified bool
    );
    insert into public.account(id, is_verified)
    values
        (1, true),
        (2, true),
        (3, false);
    savepoint a;
    -- Filter by Int
    select jsonb_pretty(
        graphql.resolve($$
            {
              accountCollection(filter: {id: {eq: 2}}) {
                edges {
                  node {
                    id
                  }
                }
              }
            }
        $$)
    );
          jsonb_pretty           
---------------------------------
 {                              +
     "data": {                  +
         "accountCollection": { +
             "edges": [         +
                 {              +
                     "node": {  +
                         "id": 2+
                     }          +
                 }              +
             ]                  +
         }                      +
     }                          +
 }
(1 row)

    rollback to savepoint a;
    -- Filter by Int and bool. has match
    select jsonb_pretty(
        graphql.resolve($$
            {
              accountCollection(filter: {id: {eq: 2}, isVerified: {eq: true}}) {
                edges {
                  node {
                    id
                  }
                }
              }
            }
        $$)
    );
          jsonb_pretty           
---------------------------------
 {                              +
     "data": {                  +
         "accountCollection": { +
             "edges": [         +
                 {              +
                     "node": {  +
                         "id": 2+
                     }          +
                 }              +
             ]                  +
         }                      +
     }                          +
 }
(1 row)

    rollback to savepoint a;
    -- Filter by Int and bool. no match
    select jsonb_pretty(
        graphql.resolve($$
            {
              accountCollection(filter: {id: {eq: 2}, isVerified: {eq: false}}) {
                edges {
                  node {
                    id
                  }
                }
              }
            }
        $$)
    );
          jsonb_pretty          
--------------------------------
 {                             +
     "data": {                 +
         "accountCollection": {+
             "edges": [        +
             ]                 +
         }                     +
     }                         +
 }
(1 row)

    rollback to savepoint a;
    -- Filter is null should have no effect
    select jsonb_pretty(
        graphql.resolve($$
            {
              accountCollection(filter: null) {
                edges {
                  node {
                    id
                  }
                }
              }
            }
        $$)
    );
          jsonb_pretty           
---------------------------------
 {                              +
     "data": {                  +
         "accountCollection": { +
             "edges": [         +
                 {              +
                     "node": {  +
                         "id": 1+
                     }          +
                 },             +
                 {              +
                     "node": {  +
                         "id": 2+
                     }          +
                 },             +
                 {              +
                     "node": {  +
                         "id": 3+
                     }          +
                 }              +
             ]                  +
         }                      +
     }                          +
 }
(1 row)

    -- neq
    select graphql.resolve($${accountCollection(filter: {id: {neq: 2}}) { edges { node { id } } }}$$);
                                        resolve                                         
----------------------------------------------------------------------------------------
 {"data": {"accountCollection": {"edges": [{"node": {"id": 1}}, {"node": {"id": 3}}]}}}
(1 row)

    rollback to savepoint a;
    -- lt
    select graphql.resolve($${accountCollection(filter: {id: {lt: 2}}) { edges { node { id } } }}$$);
                              resolve                              
-------------------------------------------------------------------
 {"data": {"accountCollection": {"edges": [{"node": {"id": 1}}]}}}
(1 row)

    rollback to savepoint a;
    -- lte
    select graphql.resolve($${accountCollection(filter: {id: {lte: 2}}) { edges { node { id } } }}$$);
                                        resolve                                         
----------------------------------------------------------------------------------------
 {"data": {"accountCollection": {"edges": [{"node": {"id": 1}}, {"node": {"id": 2}}]}}}
(1 row)

    rollback to savepoint a;
    -- gte
    select graphql.resolve($${accountCollection(filter: {id: {gte: 2}}) { edges { node { id } } }}$$);
                                        resolve                                         
----------------------------------------------------------------------------------------
 {"data": {"accountCollection": {"edges": [{"node": {"id": 2}}, {"node": {"id": 3}}]}}}
(1 row)

    rollback to savepoint a;
    -- gt
    select graphql.resolve($${accountCollection(filter: {id: {gt: 2}}) { edges { node { id } } }}$$);
                              resolve                              
-------------------------------------------------------------------
 {"data": {"accountCollection": {"edges": [{"node": {"id": 3}}]}}}
(1 row)

    rollback to savepoint a;
    -- Variable: Int
    select jsonb_pretty(
        graphql.resolve($$
           query AccountsFiltered($filt: Int!)
           {
             accountCollection(filter: {id: {eq: $filt}}) {
               edges {
                 node{
                   id
                 }
               }
             }
           }
        $$,
        variables:= '{"filt": 1}'
      )
    );
          jsonb_pretty           
---------------------------------
 {                              +
     "data": {                  +
         "accountCollection": { +
             "edges": [         +
                 {              +
                     "node": {  +
                         "id": 1+
                     }          +
                 }              +
             ]                  +
         }                      +
     }                          +
 }
(1 row)

    rollback to savepoint a;
    -- Variable: IntFilter
    select jsonb_pretty(
        graphql.resolve($$
           query AccountsFiltered($ifilt: IntFilter!)
           {
             accountCollection(filter: {id: $ifilt}) {
               edges {
                 node{
                   id
                 }
               }
             }
           }
        $$,
        variables:= '{"ifilt": {"eq": 3}}'
      )
    );
          jsonb_pretty           
---------------------------------
 {                              +
     "data": {                  +
         "accountCollection": { +
             "edges": [         +
                 {              +
                     "node": {  +
                         "id": 3+
                     }          +
                 }              +
             ]                  +
         }                      +
     }                          +
 }
(1 row)

    rollback to savepoint a;
    -- Variable: AccountFilter, single col
    select jsonb_pretty(
        graphql.resolve($$
           query AccountsFiltered($afilt: AccountFilter!)
           {
             accountCollection(filter: $afilt) {
               edges {
                 node{
                   id
                 }
               }
             }
           }
        $$,
        variables:= '{"afilt": {"id": {"eq": 2}} }'
      )
    );
          jsonb_pretty           
---------------------------------
 {                              +
     "data": {                  +
         "accountCollection": { +
             "edges": [         +
                 {              +
                     "node": {  +
                         "id": 2+
                     }          +
                 }              +
             ]                  +
         }                      +
     }                          +
 }
(1 row)

    rollback to savepoint a;
    -- Variable: AccountFilter, multi col match
    select jsonb_pretty(
        graphql.resolve($$
           query AccountsFiltered($afilt: AccountFilter!)
           {
             accountCollection(filter: $afilt) {
               edges {
                 node{
                   id
                 }
               }
             }
           }
        $$,
        variables:= '{"afilt": {"id": {"eq": 2}, "isVerified": {"eq": true}} }'
      )
    );
          jsonb_pretty           
---------------------------------
 {                              +
     "data": {                  +
         "accountCollection": { +
             "edges": [         +
                 {              +
                     "node": {  +
                         "id": 2+
                     }          +
                 }              +
             ]                  +
         }                      +
     }                          +
 }
(1 row)

    rollback to savepoint a;
    -- Variable: AccountFilter, multi col no match
    select jsonb_pretty(
        graphql.resolve($$
           query AccountsFiltered($afilt: AccountFilter!)
           {
             accountCollection(filter: $afilt) {
               edges {
                 node{
                   id
                 }
               }
             }
           }
        $$,
        variables:= '{"afilt": {"id": {"eq": 2}, "isVerified": {"eq": false}} }'
      )
    );
          jsonb_pretty          
--------------------------------
 {                             +
     "data": {                 +
         "accountCollection": {+
             "edges": [        +
             ]                 +
         }                     +
     }                         +
 }
(1 row)

    rollback to savepoint a;
    -- Variable: AccountFilter, invalid field name
    select jsonb_pretty(
        graphql.resolve($$
           query AccountsFiltered($afilt: AccountFilter!)
           {
             accountCollection(filter: $afilt) {
               edges {
                 node{
                   id
                 }
               }
             }
           }
        $$,
        variables:= '{"afilt": {"dne_id": 2} }'
      )
    );
                 jsonb_pretty                  
-----------------------------------------------
 {                                            +
     "data": null,                            +
     "errors": [                              +
         {                                    +
             "message": "Invalid filter field"+
         }                                    +
     ]                                        +
 }
(1 row)

    rollback to savepoint a;
    -- Variable: AccountFilter, invalid IntFilter
    select jsonb_pretty(
        graphql.resolve($$
           query AccountsFiltered($afilt: AccountFilter!)
           {
             accountCollection(filter: $afilt) {
               edges {
                 node{
                   id
                 }
               }
             }
           }
        $$,
        variables:= '{"afilt": {"id": 2} }'
      )
    );
                 jsonb_pretty                  
-----------------------------------------------
 {                                            +
     "data": null,                            +
     "errors": [                              +
         {                                    +
             "message": "Invalid filter field"+
         }                                    +
     ]                                        +
 }
(1 row)

    rollback to savepoint a;
    -- Variable: AccountFilter, invalid data type
    select jsonb_pretty(
        graphql.resolve($$
           query AccountsFiltered($afilt: AccountFilter!)
           {
             accountCollection(filter: $afilt) {
               edges {
                 node{
                   id
                 }
               }
             }
           }
        $$,
        variables:= '{"afilt": {"id": {"eq": true}} }'
      )
    );
                               jsonb_pretty                               
--------------------------------------------------------------------------
 {                                                                       +
     "data": null,                                                       +
     "errors": [                                                         +
         {                                                               +
             "message": "invalid input syntax for type integer: \"true\""+
         }                                                               +
     ]                                                                   +
 }
(1 row)

    rollback to savepoint a;
    -- Variable: AccountFilter, null does not apply any filters
    select jsonb_pretty(
        graphql.resolve($$
           query AccountsFiltered($afilt: AccountFilter!)
           {
             accountCollection(filter: $afilt) {
               edges {
                 node{
                   id
                 }
               }
             }
           }
        $$,
        variables:= '{"afilt": null }'
      )
    );
          jsonb_pretty           
---------------------------------
 {                              +
     "data": {                  +
         "accountCollection": { +
             "edges": [         +
                 {              +
                     "node": {  +
                         "id": 1+
                     }          +
                 },             +
                 {              +
                     "node": {  +
                         "id": 2+
                     }          +
                 },             +
                 {              +
                     "node": {  +
                         "id": 3+
                     }          +
                 }              +
             ]                  +
         }                      +
     }                          +
 }
(1 row)

    rollback to savepoint a;
rollback;
